onboarding/mlsys/1-backprop.md

```diff
@@ -0,0 +1,414 @@
+accurate predictions for a set of inputs?"
+
+You're going to write code to solve this exact problem for one particularly
+simple model.
+
+Afterwards, you will understand of basic concepts like gradient descent and
+batch size.  You will also understand how those basic concepts translate into
+some of Determined's advanced features, like gradient aggregation,
+data-parallel distributed training, and model-parallel distributed training.
+
+## The Model
+
+Our model is going to be be a simple linear model: `ypred = m*x` (`pred` is
+short for "predicted"). In statistics they would say there is no bias term, or
+in algebra they would there's no constant term.  Just a stright line through
+the origin.
+
+## The Data Set
+
+Our made-up dataset has 4 data points:
+
+- `(x=1, ylabel=1)`
+- `(x=2, ylabel=2)`
+- `(x=3, ylabel=3)`
+- `(x=4, ylabel=4)`
+
+## The Math Problem
+
+What is the best value of `m` to make our model (`ypred = m*x`) most-accurately
+predict our dataset?
+
+### Q1:
+
+The answer is `m=1`.  Looking at the dataset, explain why `m=1` is correct
+before proceeding with further math.
+
+## What does "most-accurately" mean, really?
+
+For this model and dataset, there is an exact answer (`m`=1), but usually there
+isn't.  So usually instead of seeking an exact answer, we are really seeking to
+minimize a function, called the "loss", that mathematically describes the
+goodness-of-fit between our model and the data.  There are lots of loss
+functions, but we're going to use the least-squares error function, sometimes
+abbreviated LSE.  It is defined as the sum of the squares of the error between
+our labels (the `y` values of our dataset) and our predictions (the output `y`
+of our `y=m*x` model).
+
+So for data points `(x0, ylabel0)`, `(x1, ylabel1)` and corresponding
+predictions `ypred0`, `ypred1`, you would calculate the LSE as:
+
+    loss = (ylabel0 - ypred0)**2 + (ylabel1 - ypred1)**2
+
+### Q2:
+
+Complete the following table, calculate the loss for each point of our dataset
+if we set `m=2` in our model:
+
+|  x  | ylabel | ypred | `(ylabel - ypred)**2` |
+| --- | ------ | ----- | --------------------- |
+|  1  |   1    |       |                       |
+|  2  |   2    |       |                       |
+|  3  |   3    |       |                       |
+|  4  |   4    |       |                       |
+
+Write a function for calculating the LSE loss for a single `(x, ylabel)` data
+pair.  It should take `m` and a single `(x, ylabel)` pair as inputs and return
+the loss.
+
+Then write a function for calculating the LSE loss for a whole dataset.  It
+should take `m` and a list of `(x, ylabel)` pairs and return the LSE loss.
+
+Is it possible to write the second function by making multiple calls to the
+first function and combining the results?  Why or why not?
+
+## Minimizing the loss
+
+We have the loss, but what do we do to minimize it?  The answer is called
```

Comment:
Nit: more accurately, there exists many different strategies for minimizing the loss, but one of the simplest and most prevalent methods in DL is gradient descent